packet-flow-based iptables proxy tests

danwinship opened the pull request:
#### What this PR does / why we need it:
Spinoff from #106497.

This adds a new framework for creating assertions like "given this set of iptables rules, a packet from 10.180.0.2 to 172.30.0.41:80 will get NATted to 10.180.0.1:80 without being masqueraded". This will make it easier to confirm that later refactorings don't break anything. (eg, that moving the masquerade rules around doesn't change which packets do and don't get masqueraded).

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:
none

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig network
/priority important-soon
/assign @thockin

danwinship committed:
proxy/iptables: add a general test of internal vs external traffic

Add TestInternalExternalMasquerade, which tests whether various
packets are considered internal or external for purposes of traffic
policy, and whether they get masqueraded, with and without
--masquerade-all, with and without a working LocalTrafficDetector.
(This extends and replaces the old TestMasqueradeAll.)

danwinship commented:
/retest

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -419,17 +419,44 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
-			inRightTable = (line == "*"+table)
+// parseIPTablesData takes iptables-save output and returns a map of table name to array of lines.
+func parseIPTablesData(ruleData string) (map[string][]string, error) {
+	rawTables := strings.Split(strings.TrimPrefix(ruleData, "\n"), "COMMIT\n")
```
ok, this creates an array from
```
*filter
COMMIT
*nat
:KUBE-SERVICES - [0:0]
:KUBE-SVC-XPGD46QRK7WJZT7O - [0:0]
-A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 172.30.0.41 --dport 80 -j KUBE-SVC-XPGD46QRK7WJZT7O
COMMIT
```

with 2 elements
```
element 0: *filter

element 1: *nat
:KUBE-SERVICES - [0:0]
:KUBE-SVC-XPGD46QRK7WJZT7O - [0:0]
-A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 172.30.0.41 --dport 80 -j KUBE-SVC-XPGD46QRK7WJZT7O
```

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -419,17 +419,44 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
+// parseIPTablesData takes iptables-save output and returns a map of table name to array of lines.
+func parseIPTablesData(ruleData string) (map[string][]string, error) {
+	rawTables := strings.Split(strings.TrimPrefix(ruleData, "\n"), "COMMIT\n")
+	nTables := len(rawTables) - 1
```
this is because the last element is always empty right?

aojea replied:
ah, the next `if` condition checks that  exactly

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -419,17 +419,44 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
+
+	tables := make(map[string][]string, nTables)
+	for i, table := range rawTables[:nTables] {
+		lines := strings.Split(strings.Trim(table, "\n"), "\n")
```
interesting how complex is parsing this :smile:  https://go.dev/play/p/hMn54GBM5O7

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -464,60 +491,128 @@ func moveMatchingLines(pattern string, input, output []string) ([]string, []stri
-	createdChains := sets.NewString(findAllMatches(lines, `^:([^ ]*)`)...)
-	filledChains := sets.NewString(findAllMatches(lines, `-A ([^ ]*)`)...)
+	for tableName, lines := range tables {
+		createdChains := sets.NewString(findAllMatches(lines, `^:([^ ]*)`)...)
```
this gets `:KUBE-SERVICES`

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -464,60 +491,128 @@ func moveMatchingLines(pattern string, input, output []string) ([]string, []stri
-	filledChains := sets.NewString(findAllMatches(lines, `-A ([^ ]*)`)...)
+	for tableName, lines := range tables {
+		createdChains := sets.NewString(findAllMatches(lines, `^:([^ ]*)`)...)
+		filledChains := sets.NewString(findAllMatches(lines, `-A ([^ ]*)`)...)
```
and this `-A KUBE-SERVICES -m comment`

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -464,60 +491,128 @@ func moveMatchingLines(pattern string, input, output []string) ([]string, []stri
+		jumpedChains := sets.NewString(findAllMatches(lines, `-j ([^ ]*)`)...)
+		// Ignore jumps to built-in chains
+		jumpedChains.Delete("ACCEPT", "REJECT", "DROP", "MARK", "RETURN", "DNAT", "SNAT", "MASQUERADE")
+		// KubeMarkDropChain is created by kubelet, not kube-proxy
+		jumpedChains.Delete(string(KubeMarkDropChain))
```
you have a kep about this

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -419,17 +419,44 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
-		if line[0] == '*' {
-			inRightTable = (line == "*"+table)
+// parseIPTablesData takes iptables-save output and returns a map of table name to array of lines.
+func parseIPTablesData(ruleData string) (map[string][]string, error) {
```
this is implicitly tested but ideally we can have an unit test just to cover it against refactors

aojea replied:
I say this because it is the core of all the other functions and a mistake here can be carried over and hard to notice, I think that for new test we are mostly copy and pasting current test assuming current content is right

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -464,60 +491,128 @@ func moveMatchingLines(pattern string, input, output []string) ([]string, []stri
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			err := checkIPTablesRuleJumps(tc.input)
+			if err == nil {
```
should we add a test case for the good case?

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -579,12 +671,22 @@ func assertIPTablesRulesEqual(t *testing.T, expected, result string) {
 // rules in expected, ignoring irrelevant ordering differences.
 func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
-	expected, err := sortIPTablesRules(expected)
+	err := checkIPTablesRuleJumps(expected)
+	if err != nil {
+		t.Fatalf("%s", err)
+	}
```
don't we have to check `checkIPTablesRuleJumps(result)` too ? 
we are asserting that expected and result are not equal so we can't infer from this that result rules are correct

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1098,7 +1098,7 @@ func (proxier *Proxier) syncProxyRules() {
 
 		svcXlbChain := svcInfo.serviceLBChainName
-		if svcInfo.NodeLocalExternal() {
+		if hasEndpoints && svcInfo.NodeLocalExternal() {
```
:100:

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -820,16 +885,16 @@ COMMIT
 -A KUBE-NODEPORTS -m comment --comment ns2/svc2:p80 -m tcp -p tcp --dport 3001 -s 127.0.0.0/8 -j KUBE-MARK-MASQ
 -A KUBE-NODEPORTS -m comment --comment ns2/svc2:p80 -m tcp -p tcp --dport 3001 -j KUBE-XLB-GNZBNJ2PO5MGZ6GT
--A KUBE-NODEPORTS -m comment --comment ns3/svc3:p80 -m tcp -p tcp --dport 3001 -j KUBE-SVC-X27LE4BHSL4DOUIK
+-A KUBE-NODEPORTS -m comment --comment ns3/svc3:p80 -m tcp -p tcp --dport 3002 -j KUBE-SVC-X27LE4BHSL4DOUIK
```
:upside_down_face:

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -390,7 +402,21 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
+	networkInterfacer := utilproxytest.NewFakeNetwork()
+	itf := net.Interface{Index: 0, MTU: 0, Name: "lo", HardwareAddr: nil, Flags: 0}
+	addrs := []net.Addr{
+		&net.IPNet{IP: netutils.ParseIPSloppy("127.0.0.1"), Mask: net.CIDRMask(16, 32)},
```
nit  8,32
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
```

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1111,6 +1113,258 @@ COMMIT
+// destIP, and destPort. nodeIP is the local node IP (for rules matching "LOCAL").
+// endpointProbability is the probability string to match for an endpoint jump (or "" to
+// match the final, fallback jump)
+func ruleMatches(rule, sourceIP, destIP, destPort, nodeIP, endpointProbability string) bool {
```
ok, I always wanted to do this, I think that this also has to be its own tool :smile: 
/cc @thockin

danwinship replied:
I had thought about putting it in `pkg/util/iptables`, but it's not fully generic...

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -4957,17 +5075,28 @@ COMMIT
 			fp.OnEndpointSliceAdd(testcase.endpointslice)
 			fp.syncProxyRules()
-			t.Log(fp.iptablesData.String())
-			assertIPTablesRulesEqual(t, testcase.expectedIPTables, fp.iptablesData.String())
```
I feel we loose coverage here, since the test will pass with the rules that are not exercised by the trace, right?

danwinship replied:
We're not losing signal, we're losing noise. The `assertIPTablesRulesEqual` tests were asserting too much; they made sure that the things we cared about were right, but they also made sure that a lot of things we _didn't_ care about were exactly one particular way, even though we didn't care whether they were that way or not. And this made it really really hard to figure out whether we were actually asserting what we meant to be asserting.

eg, https://github.com/kubernetes/kubernetes/pull/106497/commits/449a8d1f482622c46b87d3f9167afa6a89ee2cef is a major reorganization of the iptables rules, and you can see in the diff to `TestOverallIPTablesRulesWithMultipleServices` that a whole bunch of rules get changed as a result... But what do those changes _mean_? It's really hard to say.

But looking at the rest of the diff, you can see that even though there are a ton of changes to the generated rules, the actual observed behavior only changes in 4 of the trace tests, 2 of which had comments pointing out that they were clearly broken before. (And the other 2 cover the same cases and were also wrong, though they didn't have comments pointing that out.)

And most importantly, the current version of that commit is different from [the original version](https://github.com/kubernetes/kubernetes/commit/34d73d88d2b8797e009b94285bce16ecdcfdeab4), which turned out to be broken. In theory, you should be able to look at the changes to the `assertIPTablesRulesEqual` tests in that commit and tell that the changes are wrong. But neither Tim nor I was able to do that. Whereas once I added the trace tests, it was clear that the new rules were broken:

```
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/node_to_clusterIP (0.00s)
        rules_test.go:1014: wrong masq: expected true got false
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/external_to_clusterIP (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.0.2:80" got "DROP"
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/node_to_nodePort (0.00s)
        rules_test.go:1014: wrong masq: expected true got false
    --- FAIL: TestOverallIPTablesRulesWithMultipleServices/node_to_externalIP (0.00s)
        rules_test.go:1014: wrong masq: expected true got false
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/cluster_IP_to_non-local_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.0.1:80" got "10.180.2.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/cluster_IP_to_local_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.2.1:80" got ""
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/external_to_NodePort (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.2.1:80" got "10.180.0.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/external_to_NodePort,_no_second_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "" got "10.180.2.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/pod_to_NodePort (0.00s)
        rules_test.go:1011: wrong output: expected "10.180.2.1:80" got "10.180.0.1:80"
    --- FAIL: TestOnlyLocalNodePortsNoClusterCIDR/pod_to_NodePort,_no_second_endpoint (0.00s)
        rules_test.go:1011: wrong output: expected "" got "10.180.2.1:80"
...
```

aojea replied:
ic, so you did this based on your experience and I was judging from the top of my head, I retire my comment then since seems to be based on a false assumption

thockin replied:
To be fair, I didn't TRY to coninvce myself it was right, but your point stands.

I think the right answer is a combination of small tests which target specific functionality and use the trace *and* dump-compare approaches and larger tests which just use the traces.

Like, concretely, the comments ARE part of the tested surface, but the traces throw them out.  I am sure there are other aspects that are similar.

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1111,6 +1113,258 @@ COMMIT
+func ruleMatches(rule, sourceIP, destIP, destPort, nodeIP, endpointProbability string) bool {
+	var match []string
+
+	rule = commentRegex.ReplaceAllString(rule, "")
```
note: remove comments from rule

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1111,6 +1113,258 @@ COMMIT
+var jumpToEndpointRegex = regexp.MustCompile(" -j KUBE-SEP-")
+
+var noMatchModules = sets.NewString("conntrack", "mark")
+var knownModules = sets.NewString("addrtype", "recent", "statistic", "tcp")
```
he, we don't have any rule using `udp` or `sctp`

danwinship replied:
that's what I was saying about this not being fully generic

aojea commented:
- LGTM to commits 1-5

- commit 6,7 introduces a new "packet-flow-tracer" that I really find very useful and like it, also to have as an external tool but ...
I, personally, prefer to use it as an addition, but not for replacing current tests. I couldn't do a review in detail, but it scares me a bit the amount of regex logic it contains, because it is easy to have subtle bugs using regexes. 
Also, I found useful to have a full dump of the iptables, because it can help to detect that we don't add new rules that may go unnoticed if we only check on the packet tracing

/assign @thockin

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1111,6 +1113,258 @@ COMMIT
+	// Match local/non-local
+	match = localRegex.FindStringSubmatch(rule)
+	if match != nil {
+		sourceIsLocal := sourceIP == nodeIP || sourceIP == "127.0.0.1"
```
is `::1` required?

danwinship replied:
No, nodeports (and NAT in general) don't work with `::1`.

I think what actually happens is that the rule matches but then the `-j DNAT` fails to actually change the destination address. But pretending the rule doesn't match has the same effect.

danwinship replied:
(also, we don't use IPv6 in any of the current tests. "not fully generic")

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1367,6 +1369,258 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+	if err != nil {
+		panic("bad iptables ruleData: " + err.Error())
+	}
+
```
I think that you should add the `checkIPTablesRuleJumps` as a safeguard so people that adds a test wrong don't create an endless loop

danwinship replied:
`checkIPTablesRuleJumps` does not check that. It checks for:
- jumps to non-created chains (which would cause the e2e tests to fail too because `iptables-restore` won't let you do that, but this potentially catches it sooner)
- jumps to empty chains (which is a theoretically valid thing to do, and so would not cause `iptables-restore` failures, but which happens to be something we don't _actually_ do, so we know it's a bug if it happens)
- chains that never get jumped to (which again is theoretically valid, but not expected in our code)

I could modify `tracePacket` to keep track of the chains it passes through and notice if it ends up in a loop, but I'm not sure that's worthwhile? We never had any tests for this in the past and never had any problems... (Tim brought up the possibility of loops in the other PR, but that was a case where the code was (non-obviously) creating a SRV-to-XLB jump when we had a localDetector and an XLB-to-SRV jump when we didn't have a localDetector, so there would never have actually been any output where a loop was possible.)

danwinship commented:
/retest-required

danwinship committed:
proxy/iptables: add packet-flow-based tests of iptables rules

Add a new framework for testing out how particular packets would be
handled by a given set of iptables rules. (eg, "assert that a packet
from 10.180.0.2 to 172.30.0.41:80 gets NATted to 10.180.0.1:80 without
being masqueraded"). Add tests using this to all of the existing unit
tests.

This makes it easier to tell whether a given code change has any
effect on behavior, without having to carefully examine the diffs to
the generated iptables rules.

danwinship commented:
@aojea ok I redid this so that now we leave all of the `assertIPTablesRulesEqual` in. So now this _just_ adds tests, no removals. Want to merge it?

danwinship commented:
/retest-required

danwinship commented:
/retest-required

danwinship commented:
/retest-required

danwinship commented:
/retest-required

danwinship committed:
proxy/iptables: remove redundant test

We originally had one HealthCheckNodePort test that used
assertIPTablesRulesEqual() and one that didn't, but later I went
through and made all the tests use assertIPTablesRulesEqual() and
didn't notice that this resulted in there now being two
nearly-identical HealthCheckNodePort tests.

danwinship committed:
proxy/iptables: fix up some test case comments

danwinship commented:
/retest-required

thockin reviewed (commented):
If you bust "remove redundant test" into a PR, I'll approve it ASAP and get it off the stack.

thockin reviewed (commented):
ditto "fix up some test case comments"

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+			tracer.markDrop = true
+			continue
+
+		case "ACCEPT", "REJECT":
```
comment these are only valid in filter, and DNAT ony in nat ?

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+	// happens in nat:POSTROUTING is that the masquerade mark gets turned into actual
+	// masquerading.
+
+	tracer.runChain("nat", "KUBE-SERVICES", sourceIP, destIP, destPort)
```
comment that the jump from PREROUTING to KUBE-SERVICES is configured outside of this logic, so that's why we're manually doing it here

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+	tracer.runChain("filter", "KUBE-NODEPORTS", sourceIP, destIP, destPort)
+
+	return tracer.matches, strings.Join(tracer.outputs, ", "), tracer.markMasq
+}
```
The above was surprisingly coherent - well done.  Now that I reach the bottom, I feel like we need a suite of tests to prove that tracePacket() works against known inputs - is that not needed for some reason I am missing?

danwinship replied:
added

thockin reviewed (commented):
This turned out not as scary as I feared

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -374,6 +374,8 @@ func TestDeleteEndpointConnectionsIPv6(t *testing.T) {
 
 const testHostname = "test-hostname"
```
hm, what do you mean?

One place I was inconsistent is whether the test cases used constant strings

```
		output:   "10.180.0.1:80",
```

or generated them from local variables:

```
		output:   fmt.Sprintf("%s:%d", epIP, svcPort),
```

Some of the test functions had local variables to use and some didn't, so I just followed the lead of the existing code in each function. But I never decided which approach I liked better...

danwinship replied:
> E.g. could we do something like:

we already do; this was fixed in https://github.com/kubernetes/kubernetes/pull/108251/commits/d74df127e9cfc3d437311151a96a36bedcc176f2 as part of #108251 (inspired by [your comment here](https://github.com/kubernetes/kubernetes/pull/106497#discussion_r772690791). (And it's documented in the lines immediately above the line you commented on.)

They aren't quite perfect because I was (originally) trying to avoid illegally using "real" IPs, but there were only so many easily-differentiated private/reserved IP ranges to pick from. So I made node IPs be 192.168.0.x and ExternalIPs be 192.168.99.x (which kind of makes sense because in many configurations, ExternalIPs are on the same subnet as the nodes). And "cluster-external and allowed" and "cluster-external but blocked" only differ in 1 bit, but that doesn't really matter because we only make that distinction in a few places and the test names always make it clear what's going on...

But then I ran out of distinguishable ranges and ended up using 1.2.3.4, 5.6.7.8, and 9.10.11.12 for load balancer IPs, violating my "no real IPs" rule. So I guess we could use other real IP ranges to better differentiate the other ranges.

Or we could just switch to IPv6 and encode the meanings in leetspeek (`70ad:ba7a:ce00::1`)...

thockin replied:
I hate that I love `10ad:ba1a:ce00::1`, but let's not :)

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+var destPortRegex = regexp.MustCompile(" --dport ([^ ]*) ")
+var sourceIPRegex = regexp.MustCompile("(!)? -s ([^ ]*) ")
+var affinityRegex = regexp.MustCompile(" --rcheck ")
+var jumpRegex = regexp.MustCompile("(--probability.*)? -j ([^ ]*)( --to-destination (.*))?$")
```
In this case it couldn't be reordered. (The overall syntax is `[match clauses...] [-j target [target options...]]`.)

But if you could reorder the clauses, and someone did, then the patterns wouldn't match any more and a bunch of tests would fail...

thockin replied:
ehhhhh ok.  I'm a big fan of comments to leave these sorts of clues for future-me :)

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+	}
+
+	// Match destination IP/port.
+	// destIPRegex = /(!)? -d ([^ ]*) /
```
It was an attempt to clarify what `matches[1]`, `matches[2]`, etc meant. But if we switch to a more procedure-based parsing then that wouldn't matter

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+}
+
+// ruleMatches checks if the given iptables rule matches (at least probabilistically) a
+// packet with the given sourceIP, destIP, and destPort.
```
oh, hm, that wasn't really intentional...

thockin replied:
a comment would suffice

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+		if !tracer.ruleMatches(rule, sourceIP, destIP, destPort) {
+			continue
+		}
+		tracer.matches = append(tracer.matches, rule)
```
(debugging... it's mentioned in the struct definition)

danwinship replied:
oops. it _was_ mentioned in the struct definition in an earlier version of the patch...

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+		if noMatchModules.Has(moduleMatches[1]) {
+			return false
+		} else if !knownModules.Has(moduleMatches[1]) {
+			klog.ErrorS(nil, "Rule uses unknown iptables module; assuming no match", "rule", rule, "module", moduleMatches[1])
```
I made this panic, though it occured to me later that this isn't _really_ saving us because there are other rules that are allowed without any `-m` flag which we don't make any effort to understand (eg `-i` / `-o`) and for the `knownModules` we don't actually try to parse _all_ of the rules they define (eg, `-m tcp --sport XXX`)

danwinship commented:
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/remove-sig node
/remove-sig storage
/remove-sig testing
/triage accepted

danwinship commented:
[This](https://github.com/kubernetes/kubernetes/compare/817ff6a3d4c7122c4aeca2a1a3f43fcc80222318..8c60d1100097a8c8811e3cfc7817d28788a76e88#diff-92f41de420a097c0419a3cb53cf176a459a742c39f3fa482ab48b1c2fa6fd084) shows the rebase for #108691 and the only change to the existing flow tests is that in `TestEndpointSliceWithTerminatingEndpointsTrafficPolicyLocal`, `"ProxyTerminatingEndpoints enabled, terminating endpoints on remote node"`, we used to expect that pod-to-clusterIP would fail because there were no non-terminating endpoints, but now it succeeds because we use the terminating endpoints for the internal Cluster traffic as well.

The flow tests in `TestEndpointSliceWithTerminatingEndpointsTrafficPolicyCluster` are new, based on the ones from `...Local` but with appropriate changes to the results.

There are no flow test changes as a result of #109060, but I added in `getLine()` usage like in the `assertIPTablesRulesEqual()` tests.

pacoxu commented:
/retest

dims commented:
/assign @aojea

jyotimahapatra commented:
@danwinship 👋 1.24 bug triage team checking in. The 1.24 test freeze is at Thursday 01:00 UTC(Approximately 1 day remaining). Could you please get lgtm by then. 
Also for our familiarity, what is the risk of pushing this to milestone 1.25?

danwinship committed:
proxy/iptables: tweak sortIPTablesRules a bit

It's confusing to have the "this must be the last rule" get sorted to
be the first rule in KUBE-SERVICES...

danwinship commented:
@jyotimahapatra it's not urgent; it's just unit test improvements. I'll probably backport it but it can wait for 1.25 at this point.

jyotimahapatra commented:
/milestone 1.25

jyotimahapatra commented:
/milestone v1.25

thockin reviewed (commented):
Looks great!   Just some nits for now.

I can't promise to have fully comprehended every single test case, but I did try.  I feel like a good followup might be to write out the matrix of all the cases we think should have coverage and then make sure test funcs match that.  I strongly suspect there's dumplication here as well as totally missing cases.

e.g.

src: [pod, node, external]
dst: [service, same-node-nodeport, other-node-nodeport, LBIP, ExternalIP, HCNP]
itp: [local, cluster]
etp: [local, cluster]
has-endpoints: [true, false]
has-local-detector: [true, false]
lb-source-range-allows-src: [true, false]

I am sure I am missing some.  I have no idea what the sanest, most future-safe way to structure the code to cover all combinations is.  For Service REST I wrote a program to generate tens of thousands of distinct cases, then manually culled some axes that nullified others (like testing cluster IP doersn't care about lb-source-range-allows-src).

I suspect in building bigger tables, we may want to set up the tracer one time and "reset" the outputs between uses.

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1369,6 +1375,457 @@ func assertIPTablesRulesNotEqual(t *testing.T, line int, expected, result string
+	// happens in nat:POSTROUTING is that the masquerade mark gets turned into actual
+	// masquerading.
+
+	// FIXME: we ought to be able to say
```
funny you should mention that...

I'm working on a PR to [minimize iptables-restore input by not re-syncing service chains that didn't change between syncs](https://github.com/danwinship/kubernetes/commits/minimize-iptables-changes) (WIP) but it turned out to be impossible to [unit-test properly](https://github.com/danwinship/kubernetes/commit/33d5965e1e9d5a979c54eceecf4dbc5f43915dd0#diff-92f41de420a097c0419a3cb53cf176a459a742c39f3fa482ab48b1c2fa6fd084R6071) with our broken FakeIPTables, so I had to fix FakeIPTables. And then given that, we _can_ do the right thing here...

But then, fixing FakeIPTables had a lot of overlap with the parsing stuff in this branch, and based on that and your earlier comments about the regexp parsing stuff, I ended up reimplementing this branch on top of that, the current state of which is [here](https://github.com/danwinship/kubernetes/commits/iptables-tests-new) (WIP) which among other things [totally rewrites the iptables rules parsing](https://github.com/danwinship/kubernetes/commit/e727638b041c45c8f6a1e24926a8ac25e76ae175#diff-4c4b323bcf1bd7d7b30dcd59613c137bc82347747cd583ef6b8873dec94b467cR28)...

So I don't know if you'd rather get this in now, or wait for that. I can re-rebase that if this merges first.

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -6833,114 +6833,574 @@ func TestEndpointSliceWithTerminatingEndpointsTrafficPolicyCluster(t *testing.T)
+			destIP:   testNodeIP,
+			destPort: 3002,
+
+			// FIXME: The short-circuit rule means we potentially send to a remote
```
I'm not sure I agree, but I am not sure I don't.

When `eTP:Cluster` we always masquerade (at least we always have in the past) , I think that's generally agreed, right?  I don't see why we would change that.

We decided local source + `eTP:Local` is a special case.  If we want to emulate up-and-out-and-back-down, masquerade would be right for external LBs (e.g. things that leave a cloud's VPC) but would be wrong for internal LBs (e.g. in the local network).

I *think* I'd rather err on the side of "errantly retain information" than "errantly lose information".

Is there some case where the masquerade in this case is semantically important?

danwinship replied:
I wasn't saying "we should masquerade in the eTP:Local case", I was saying "either the masquerading in the Cluster case is necessary (in which case Local is broken), OR the masquerading in the Cluster case is not necessary (in which case why are we doing it?)"

I was assuming we wanted to avoid masquerading any time we could avoid it, but I guess "we always masquerade all traffic to EXT in the Cluster case, so therefore we masquerade pod-to-EXT traffic" is a plausible argument...

thockin replied:
> I was assuming we wanted to avoid masquerading any time we could avoid it, but I guess "we always masquerade all traffic to EXT in the Cluster case, so therefore we masquerade pod-to-EXT traffic" is a plausible argument...

Yeah, I have always assumed that it would more confusing to avoid SNAT in some cases but not others.  Ergo "When eTP:Cluster we always masquerade" even if it isn't strictly necessary for local clients.

Plausible?

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -6833,114 +6833,574 @@ func TestEndpointSliceWithTerminatingEndpointsTrafficPolicyCluster(t *testing.T)
+			destIP:   "5.6.7.8",
+			destPort: 80,
+
+			// FIXME: The short-circuit rule means we potentially send to a remote
```
ditto

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -5852,83 +6938,594 @@ func TestEndpointSliceWithTerminatingEndpointsTrafficPolicyCluster(t *testing.T)
-	ipt := iptablestest.NewFake()
-	fp := NewFakeProxier(ipt)
-	fp.masqueradeAll = true
+func TestInternalExternalMasquerade(t *testing.T) {
```
This is a great test.  I wish we could make setup even easier, but that's for later.

danwinship commented:
changed panics to Fatalf/Errorf (depending on whether it's a test case bug or a code bug), added comments, added `TestTracePackets` with a basic set of tests against static iptables data, and... uh... fixed a bug in tracePackets revealed by the test :slightly_smiling_face: 

```
        // Process pending DNAT (which theoretically might affect REJECT/ACCEPT filter rules)
        if len(tracer.outputs) != 0 {
-               destIP = tracer.outputs[0]
+               destIP = strings.Split(tracer.outputs[0], ":")[0]
        }
```

if we tested normal endpoint output in a ruleset that also had a REJECT rule in the filter table, it would fail because we were setting `destIP` to something like `10.180.0.1:80` rather than `10.180.0.1`.

thockin reviewed (commented):
/lgtm
/approve

We can argue FIXMEs after, I know you have followups
